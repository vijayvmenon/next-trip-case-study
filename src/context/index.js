import { useContext } from 'react';
import AppContext, { Context } from './context';

export default AppContext;
export { Context };
export const useMyContext = () => useContext(Context);
