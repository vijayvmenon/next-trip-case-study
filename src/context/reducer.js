import constants from '../constants';
import ENTranslations from '../locales/en.json';
import { localStorageExists } from '../utils';

const localStorageLanguage =
  localStorageExists && localStorage.getItem('LANGUAGE');
const initialState = {
  tableExpanded: false,
  hideHeader: false,
  snackbar: { value: '', text: '' },
  language: localStorageLanguage || 'en',
  translations: localStorageLanguage
    ? constants.LANGUAGE_MAPPING[localStorageLanguage]
    : ENTranslations,
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case 'TABLE_EXPAND':
      return { ...state, tableExpanded: true };
    case 'TABLE_COLLAPSE':
      return { ...state, tableExpanded: false };
    case 'SET_SNACKBAR':
      return { ...state, snackbar: { value: action.value, text: action.text } };
    case 'HIDE_HEADER':
      return { ...state, hideHeader: true };
    case 'SHOW_HEADER':
      return { ...state, hideHeader: false };
    case 'SET_LANGUAGE':
      return {
        ...state,
        language: action.value,
        translations: constants.LANGUAGE_MAPPING[action.value],
      };
    default:
      return { ...state };
  }
}

export { initialState };
