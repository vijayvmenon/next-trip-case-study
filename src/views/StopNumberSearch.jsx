/* eslint-disable */
import React, { useState } from 'react';
import { createUseStyles } from 'react-jss';
import { useLocation } from 'wouter';
import Input from '../components/Input';
import constants from '../constants';
import { httpGet } from '../utils';
import TableWrapper from './TableWrapper';

const useStyles = createUseStyles({
  root: {
    marginTop: 24,
    width: '60%',
  },
});

export default function StopNumberSearch() {
  const classes = useStyles();
  const [inpValue, setInpValue] = useState('');
  const [tableData, setTableData] = useState(null);
  const [intrvel, setIntrvl] = useState('initial');
  const [error, setError] = useState(false);
  const [location, setLocation] = useLocation();
  console.log(location);
  const onSearch = async (value, e) => {
    if (e) {
      e.preventDefault();
    }
    if (!value) {
      return;
    }
    setLocation(`/byStop/${value}`);
    getTableData('on-search');
  };

  async function getTableData(interval) {
    setError(false);
    /* set tableData to null only if a new search is made. 
    This is to handle the useEffect run for clearing interval in <Table/> component */
    interval === 'on-search' && setTableData(null);
    setIntrvl(interval);
    const [response, error] = await httpGet(inpValue);
    if (response) {
      setTableData(response);
    }
    if (error) {
      setIntrvl('error');
      setError(true);
    }
  }

  return (
    <>
      <form className={classes.root} action="">
        <Input
          type="number"
          label={constants.ENTER_STOP_NUMBER}
          onSearch={onSearch}
          value={inpValue}
          setInput={(value) => setInpValue(value)}
        />
      </form>
      <TableWrapper
        tableData={tableData}
        rowField="departures"
        tableHeaderField="stops"
        interval={intrvel}
        getTableData={getTableData}
      />
    </>
  );
}
