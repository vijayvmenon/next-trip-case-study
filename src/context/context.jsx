import React, { useReducer, createContext } from 'react';
import PropTypes from 'prop-types';
import reducer, { initialState } from './reducer';

export const Context = createContext();

export default function AppContext({ children }) {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>
  );
}

AppContext.propTypes = {
  children: PropTypes.objectOf(PropTypes.any).isRequired,
};
