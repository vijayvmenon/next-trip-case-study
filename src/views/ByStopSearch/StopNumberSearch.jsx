import React, { useState, useEffect, useContext } from 'react';
import { createUseStyles } from 'react-jss';
import { useLocation } from 'wouter';
import Input from '../../components/UI/Input';
import { Context } from '../../context';
import { httpGet, checkIfUrlIsInByStopFormat } from '../../utils';
import TableWrapper from '../TableWrapper/TableWrapper';
import Snackbar from '../../components/UI/Snackbar';
import constants from '../../constants';

const useStyles = createUseStyles({
  root: {
    marginTop: 24,
    width: '100%',
    maxWidth: 570,
  },
});

export default function StopNumberSearch() {
  const classes = useStyles();
  const [inpValue, setInpValue] = useState('');
  const [tableData, setTableData] = useState(null);
  const {
    dispatch,
    state: {
      snackbar: { value: snackbarType = '', text = '' } = {},
      translations = {},
    } = {},
  } = useContext(Context);
  const [loader, setLoader] = useState(false);
  const [routeOnReload, setRouteOnReload] = useState('');
  const [intrvel, setIntrvl] = useState('initial');
  const [location, setLocation] = useLocation();

  async function getTableData(interval, passedValue) {
    setLoader(true);
    /* set tableData to null only if a new search is made. 
    This is to handle the useEffect run for clearing interval in <TableWrapper/> component */
    if (interval === 'on-search') {
      setTableData(null);
    }
    setIntrvl(interval);
    /* call the API on stop input being entered, when the url reloads and a 
    stop_id exists in the url or on interval call from TableWrapper */
    const [fetchResponse, fetchError] = await httpGet(
      passedValue || routeOnReload || inpValue,
    );
    if (fetchResponse) {
      setTableData(fetchResponse);
    }
    if (fetchError) {
      // do not set interval and clear previous intevals in error condition
      setIntrvl('error');
      // Open snackbar based on warning or error from the API call
      if (fetchError === 'warning') {
        dispatch({
          type: 'SET_SNACKBAR',
          value: 'warning',
          text: `${passedValue || inpValue} ${translations.NOT_VALID_STOP}`,
        });
      } else {
        dispatch({
          type: 'SET_SNACKBAR',
          value: 'error',
          text: translations.SNACKBAR_STOP_ERROR,
        });
      }
    }
    setLoader(false);
  }

  const onSearch = async (value, e) => {
    // Reset stop_id in url when user searches through the input
    setRouteOnReload('');
    if (e) {
      e.preventDefault();
    }
    if (!value) {
      return;
    }
    setLocation(`${constants.BASE_PATH}/byStop/${value}`);
  };

  /* Load departures (if url has required parameters) on Initial Load . This useEffect also loads the departures data when
  user searches something from the input. Since location is passed in the array of dependencies and we are setting 
  location in line 79, this useEffect will run and get the TableData. It is written like this to handle situation when the 
  user clicks browser back button */
  useEffect(() => {
    dispatch({ type: 'SHOW_HEADER' }); // show Header component again when redirecting from a page not found
    const url = checkIfUrlIsInByStopFormat(location);
    if (url) {
      setRouteOnReload(url);
      getTableData('on-search', url);
    } else {
      /* Handle situation where user presses browser back button and comes to the root /byStop page. We need to reset
      the interval if any are set */
      setTableData(null);
      setIntrvl('reset');
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [dispatch, location]);

  return (
    <>
      <Snackbar type={snackbarType} text={text} />
      <form className={classes.root} action="">
        <Input
          inputType="number"
          label={translations.ENTER_STOP_NUMBER}
          onSearch={onSearch}
          value={inpValue}
          setInput={(value) => setInpValue(value)}
          focusOnLoad
        />
      </form>
      <TableWrapper
        tableData={tableData}
        rowField="departures"
        tableHeaderField="stops"
        interval={intrvel}
        getTableData={getTableData}
        loader={loader}
      />
    </>
  );
}
