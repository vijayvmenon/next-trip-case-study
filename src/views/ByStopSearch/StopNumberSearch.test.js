import { render, fireEvent, screen } from '@testing-library/react';
import StopNumberSearch from './StopNumberSearch';
import renderer from 'react-test-renderer';
import { Context } from '../../context';

const dispatch = jest.fn(); // mock dispatch

describe('Component:StopNumberSearch.jsx', () => {
  test('StopNumberSearch.jsx renders correctly', () => {
    const stopNumberSearch = renderer
      .create(
        <Context.Provider value={{ dispatch }}>
          <StopNumberSearch />
        </Context.Provider>,
      )
      .toJSON();
    expect(stopNumberSearch).toMatchSnapshot();
  });
});
