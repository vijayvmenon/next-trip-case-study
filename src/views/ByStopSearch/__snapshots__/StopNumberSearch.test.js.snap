// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component:StopNumberSearch.jsx StopNumberSearch.jsx renders correctly 1`] = `
Array [
  "",
  <form
    action=""
    className="root-0-2-1"
  >
    <div
      className="root-0-2-10"
    >
      <input
        className="input-0-2-12"
        data-testid="custom-input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        required={true}
        type="text"
        value=""
      />
      <label
        className="label-0-2-13"
        htmlFor=""
      >
        
      </label>
      <div
        className="inputAdornments-0-2-16"
      >
        
        <button
          className="searchButton-0-2-19"
          data-testid="input-search-icon-button"
          onClick={[Function]}
          type="button"
        >
           
          <svg
            className="searchIcon-0-2-14"
          >
            search-icon.svg
          </svg>
        </button>
      </div>
    </div>
  </form>,
]
`;
